File: spring-ai-alibaba-data-agent-frontend/src/services/datasource.ts
/*
 * Copyright 2024-2025 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import axios from 'axios';
import { ApiResponse } from '@/services/common'

interface Datasource {
    id?: number;
    name?: string;
    type?: string;
    host?: string;
    port?: number;
    databaseName?: string;
    username?: string;
    password?: string;
    connectionUrl?: string;
    status?: string;
    testStatus?: string;
    description?: string;
    creatorId?: number;
    createTime?: string; // 使用字符串表示日期时间，格式为 "yyyy-MM-dd HH:mm:ss"
    updateTime?: string; // 使用字符串表示日期时间，格式为 "yyyy-MM-dd HH:mm:ss"
}

// 定义 AgentDatasource 接口
interface AgentDatasource {
    id?: number;
    agentId?: number;
    datasourceId?: number;
    isActive?: number | boolean;
    createTime?: string;
    updateTime?: string;
    datasource?: Datasource;
}

const API_BASE_URL = '/api/datasource';

class DatasourceService {
    // 1. 获取所有数据源列表
    async getAllDatasource(
        status?: string,
        type?: string
    ): Promise<Datasource[]> {
        const params = new URLSearchParams();
        if (status) params.append('status', status);
        if (type) params.append('type', type);

        const response = await axios.get<Datasource[]>(
            `${API_BASE_URL}${params.toString() ? `?${params.toString()}` : ''}`
        );
        return response.data;
    }

    // 2. 根据 ID 获取数据源详情
    async getDatasourceById(id: number): Promise<Datasource | null> {
        try {
            const response = await axios.get<Datasource>(`${API_BASE_URL}/${id}`);
            return response.data;
        } catch (error) {
            if (axios.isAxiosError(error) && error.response?.status === 404) {
                return null;
            }
            throw error;
        }
    }

    // 3. 创建数据源
    async createDatasource(datasource: Datasource): Promise<Datasource> {
        const response = await axios.post<Datasource>(API_BASE_URL, datasource);
        return response.data;
    }

    // 4. 更新数据源
    async updateDatasource(
        id: number,
        datasource: Datasource
    ): Promise<Datasource> {
        const response = await axios.put<Datasource>(`${API_BASE_URL}/${id}`, datasource);
        return response.data;
    }

    // 5. 删除数据源
    async deleteDatasource(id: number): Promise<ApiResponse<void>> {
        const response = await axios.delete<ApiResponse<void>>(`${API_BASE_URL}/${id}`);
        return response.data;
    }

    // 6. 测试数据源连接
    async testConnection(id: number): Promise<ApiResponse<boolean>> {
        const response = await axios.post<ApiResponse<boolean>>(`${API_BASE_URL}/${id}/test`);
        return response.data;
    }

}

export default new DatasourceService();

File: data-agent-management/src/main/java/com/alibaba/cloud/ai/dataagent/entity/Datasource.java
/*
 * Copyright 2024-2025 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.cloud.ai.dataagent.entity;

import com.fasterxml.jackson.annotation.JsonFormat;
import lombok.Builder;
import org.springframework.format.annotation.DateTimeFormat;

import java.time.LocalDateTime;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import org.springframework.util.StringUtils;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Datasource {

	private Integer id;

	private String name;

	private String type;

	private String host;

	private Integer port;

	private String databaseName;

	private String username;

	private String password;

	private String connectionUrl;

	private String status;

	private String testStatus;

	private String description;

	private Long creatorId;

	@JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
	@DateTimeFormat(pattern = "yyyy-MM-dd HH:mm:ss")
	private LocalDateTime createTime;

	@JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
	@DateTimeFormat(pattern = "yyyy-MM-dd HH:mm:ss")
	private LocalDateTime updateTime;

	/**
	 * Generate connection URL
	 */
	// todo: 改为策略模式
	public void generateConnectionUrl() {
		if (StringUtils.hasText(connectionUrl)) {
			return;
		}
		if (host != null && port != null && databaseName != null) {
			if ("mysql".equalsIgnoreCase(type)) {
				this.connectionUrl = String.format(
						"jdbc:mysql://%s:%d/%s?useUnicode=true&characterEncoding=utf-8&zeroDateTimeBehavior=convertToNull&transformedBitIsBoolean=true&allowMultiQueries=true&allowPublicKeyRetrieval=true&useSSL=false&serverTimezone=Asia/Shanghai",
						host, port, databaseName);
			}
			else if ("postgresql".equalsIgnoreCase(type)) {
				this.connectionUrl = String.format(
						"jdbc:postgresql://%s:%d/%s?useUnicode=true&characterEncoding=utf-8&useSSL=false&serverTimezone=Asia/Shanghai",
						host, port, databaseName);
			}
			else if ("h2".equalsIgnoreCase(type)) {
				this.connectionUrl = String.format(
						"jdbc:h2:mem:%s;DB_CLOSE_DELAY=-1;DATABASE_TO_LOWER=true;MODE=MySQL;DB_CLOSE_ON_EXIT=FALSE",
						databaseName);
			}
			else if ("dameng".equalsIgnoreCase(type)) {
				this.connectionUrl = String.format("jdbc:dm://%s:%d", host, port);
			}
		}
	}

	@Override
	public String toString() {
		return "Datasource{" + "id=" + id + ", name='" + name + '\'' + ", type='" + type + '\'' + ", host='" + host
				+ '\'' + ", port=" + port + ", databaseName='" + databaseName + '\'' + ", status='" + status + '\''
				+ ", testStatus='" + testStatus + '\'' + ", createTime=" + createTime + ", updateTime=" + updateTime
				+ '}';
	}

}
